pkg/utils/kubernetes/reconciler.go

```diff
@@ -0,0 +1,120 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"github.com/RHsyseng/operator-utils/pkg/resource"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strings"
+)
+
+type Finalizer interface {
+	GetName() string
+	OnFinalize(owner resource.KubernetesResource, service PlatformService) error
+}
+
+type ExtendedReconciler struct {
+	Service    PlatformService
+	Reconciler reconcile.Reconciler
+	Resource   resource.KubernetesResource
+	Finalizers map[string]Finalizer
+}
+
+func NewExtendedReconciler(service PlatformService, reconciler reconcile.Reconciler, resource resource.KubernetesResource) ExtendedReconciler {
+	return ExtendedReconciler{
+		Service:    service,
+		Reconciler: reconciler,
+		Resource:   resource,
+		Finalizers: map[string]Finalizer{},
+	}
+}
+
+func (e *ExtendedReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	instance := e.Resource.DeepCopyObject().(resource.KubernetesResource)
+	err := e.Service.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	err = e.finalizeOnDelete(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	return e.Reconciler.Reconcile(request)
+}
+
+func (e *ExtendedReconciler) RegisterFinalizer(f Finalizer) error {
```

Comment:
When is the finalizer string written to the resource in k8s? Did I lose track of it, or did it get lost in the refactor?

Reply:
The users will add finalizers by string as usual. https://github.com/ruromero/kie-cloud-operator/commit/48d456328b241c8fcc2062593df6c190ff60fed3#diff-9ecb910ca7304714ab23ba15d8cf55a5R596

Reply:
Maybe I'm confused... or is your link referring to operator-utils code that no longer exists in your recent commits?

Reply:
No, the code is directly using the controller-runtime#AddFinalizer. I think users should add finalizers in their reconcile loop agnostic to the ExtendedReconciler.

Reply:
Ah ok, that's what I missed. But why not set it if not set when finalizer is being registered? Other than automating another step, it helps avoid inconsistency too, no?

Reply:
Because you register finalizers during the initialization but you won't have to add the finalizer to the object until you need it.
In the case of the consoleLink, we add the finalizer to the CRD only when we have the route created but for the case of trial environment where no https route exists there won't be a consoleLink, therefore no need to call the finalizer.